tools/kvmexit.py

```diff
@@ -0,0 +1,293 @@
+#  ...
+#
+# @TGID: each vitual machine's pid in the user space.
+# @PID: the user space's thread of each vcpu of that virtual machine.
+# @KVM_EXIT_REASON: the reason why the vm exits.
+# @STAT: the counts of the @KVM_EXIT_REASON since the vm starts, and
+#        it is recorded every second by setting `sleep(1)`.
+#
+# REQUIRES: Linux 4.7+ (BPF_PROG_TYPE_TRACEPOINT support)
+#
+# Copyright (c) 2019 BYTEDANCE Inc.
+#
+# Author(s):
+#   Fei Li <lifei.shirley@bytedance.com>
+
+
+from __future__ import print_function
+from time import sleep, strftime
+from bcc import BPF
+import argparse
+import multiprocessing
+import signal
+import subprocess
+
+#
+# Process Arguments
+#
+
+# arg validation
+def positive_int(val):
+    try:
+        ival = int(val)
+    except ValueError:
+        raise argparse.ArgumentTypeError("must be an integer")
+
+    if ival < 0:
+        raise argparse.ArgumentTypeError("must be positive")
+    return ival
+
+def positive_nonzero_int(val):
+    ival = positive_int(val)
+    if ival == 0:
+        raise argparse.ArgumentTypeError("must be nonzero")
+    return ival
+
+# arguments
+examples = """examples:
+    ./kvmexit             # Display kvm_exit_reason and its statistics until Ctrl-C
+    ./kvmexit 5           # Display kvm_exit_reason and its statistics for 5 seconds only
+"""
+parser = argparse.ArgumentParser(
+    description="Display kvm_exit_reason and its statistics at a timed interval",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("duration", nargs="?", default=99999999,
+    type=positive_nonzero_int,
+    help="duration of trace, in seconds")
+
+args = parser.parse_args()
+duration = int(args.duration)
+
+#
+# Setup BPF
+#
+
+# load BPF program
+bpf_text = """
+#include <linux/delay.h>
+
+#define REASON_NUM 69
+#define TGID_NUM 1024
+
+struct exit_count {
+    u64 exit_ct[REASON_NUM];
+};
+BPF_PERCPU_ARRAY(init_value, struct exit_count, 1);
+BPF_TABLE("percpu_hash", u64, struct exit_count, pcpu_kvm_stat, TGID_NUM);
+
+struct cache_info {
+    u64 cache_pid_tgid;
+    struct exit_count cache_exit_ct;
+};
+BPF_PERCPU_ARRAY(pcpu_cache, struct cache_info, 1);
+
+FUNC_ENTRY {
+    u64 cur_pid_tgid = bpf_get_current_pid_tgid();
+    int zero = 0;
+    int cache_miss = 0;
+    u32 er = GET_ER;
+    if (er >= REASON_NUM) {
+        return -1;
+    }
+
+    struct exit_count *tmp_info, *initial;
+    struct cache_info *cache_p;
+    cache_p = pcpu_cache.lookup(&zero);
+    if (cache_p != NULL) {
+        if (cache_p->cache_pid_tgid == cur_pid_tgid) {
+            //a. If the cur_pid_tgid hit this physical cpu consecutively, save it to pcpu_cache
+            tmp_info = &cache_p->cache_exit_ct;
+        } else {
+            //b. If another pid_tgid matches this pcpu for the last hit, OR it is the first time to hit this physical cpu.
+            cache_miss = 1;
+        }
+    } else {
+        return -1;
+    }
```

Comment:
typically we should do

  if (cache_p == NULL)
    return 0;
  ...

to reduce nesting.

Also this caching stuff seems complicated. Will go back to this once I understand use case better.

Reply:
> typically we should do
> 
> if (cache_p == NULL)
> return 0;
> ...
> 
> to reduce nesting.
NP
> 
> Also this caching stuff seems complicated. Will go back to this once I understand use case better.
Would the `Help` and `flow` in "tools/kvmexit_example.txt" be helpful to understand the caching stuff? Hope they  work, as that's what they're supposed to..